csharp/Adapter/Microsoft.Spark.CSharp/Network/ByteBufPool.cs

```diff
@@ -110,23 +116,31 @@ public static ByteBufPool UnsafeDefault
         [MethodImpl(MethodImplOptions.Synchronized)]
         public ByteBuf Allocate()
         {
+            var trial = 0;
             ByteBuf byteBuf;
-            if (q050.Allocate(out byteBuf) || q025.Allocate(out byteBuf) ||
-                q000.Allocate(out byteBuf) || qInit.Allocate(out byteBuf) ||
-                q075.Allocate(out byteBuf))
+            do
             {
-                return byteBuf;
-            }
+                if (q050.Allocate(out byteBuf) || q025.Allocate(out byteBuf) ||
+                    q000.Allocate(out byteBuf) || qInit.Allocate(out byteBuf) ||
+                    q075.Allocate(out byteBuf))
+                {
+                    return byteBuf;
+                }
+                Thread.Sleep(0);
```

Comment:
Thread.Sleep(0) issues a pause instruction on each loop iteration, and not fall back to a more expensive wait on.